test/plug/conn_test.exs

```diff
@@ -405,6 +405,20 @@ defmodule Plug.ConnTest do
     assert get_resp_header(conn, "x-body") == ["CHUNK"]
   end
 
+  test "chunk/2 runs before_chunk callbacks" do
+    pid = self()
+    _conn =
+      conn(:get, "/foo")
+      |> register_before_chunk(fn conn, chunk ->
+        send(pid, {:before_chunk, chunk})
+        conn
+       end)
+      |> send_chunked(200)
+      |> chunk("CHUNK")
+
+    assert_received {:before_chunk, "CHUNK"}
+  end
```

Comment:
Ah, we also need to add tests for:

  * the `AlreadySentError` exception
  * the error that you raised when the state of the conn changes
  * a test where you modify the connection in one `register_before_chunk/2` callback (like adding a header or something) and you make sure that the new connection is passed to further callbacks

Reply:
I added a test for 1. For 2-3, almost all the methods which modify `Conn` raise an `AlreadySentError` since `conn.state == :chunked` fails the `state in @unsent` guard.

We could modify the `conn` manually inside the callback for the purpose of the test, but that feels a bit contrived... Thoughts?

Reply:
I think we should simply document the `conn` is read-only, as you can't really do anything useful with them. The main goal is to get a copy of any assigns or headers stored in it.

Reply:
Actually, scratch that, I can see a connection is being returned all the way up.

@feynmanliang you can set a assigns/private field, like a counter, and measure it is being bumped. That should be supported :)

Reply:
Sounds good, I've exercised that in https://github.com/elixir-plug/plug/pull/1154/files#diff-5115eb93e759386d8da8b04d13db4e24d30f2d9ece59b85fe63068236aac9dfeR427-R444